### PR TITLE
Allow forward declarations bring additional IR dependencies for modules present in metadata.

### PR DESCRIPTION
**Pull request by alexander-gorshenev**
* Eliminated global state in IR deserializer. 
* Disentangled module header creation and declaration deserialization
* Allow forward declarations bring additional IR dependencies for modules present in metadata.

### CODE REVIEW

**Comment by alexander-gorshenev** [comment]
The native counterpart is here: https://github.com/JetBrains/kotlin-native/pull/2800

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/compiler.kt [hunk 1]
```diff
@@ -142,7 +144,7 @@ fun compile(
 
     val deserializedModuleFragments = sortedImmediateDependencies.map {
         val moduleFile = File(it.klibPath, moduleHeaderFileName)
-        deserializer.deserializeIrModuleHeader(depsDescriptors.getModuleDescriptor(it), moduleFile.readBytes(), File(it.klibPath), DeserializationStrategy.ONLY_REFERENCED)
+        deserializer.DeserializeIrModuleHeader(depsDescriptors.getModuleDescriptor(it))
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/compiler.kt]: deserializer.~D~deserializeIrModuleHeader

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt [hunk 2]
```diff
@@ -294,6 +364,7 @@ abstract class KotlinIrLinker(
             ?: error("findDeserializedDeclaration: property descriptor $propertyDescriptor} is not present in propertyTable after deserialization}")
     }
 
+    // TODO: This is Native specific. Eliminate me.
```
Comment (romanart) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt]: I agree that this method should be abstract

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/IrModuleDeserializer.kt [hunk 3]
```diff
@@ -44,7 +43,7 @@ abstract class IrModuleDeserializer(
     val logger: LoggingContext,
     val builtIns: IrBuiltIns,
     val symbolTable: SymbolTable
-) : IrDeserializer {
+) {
```
Comment (romanart) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/IrModuleDeserializer.kt]: That weird that `IrModuleDeserializer` is no more `IrDeserializer` when `KotlinIrLinker` is

Comment (alexander-gorshenev) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/IrModuleDeserializer.kt]: may be I need to rename `IrDeserializer` interface to `IrLinker` already.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt (hunk 2 above)
Comment (alexander-gorshenev) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt]: I am going to eliminate the remaining native and js specific pieces from the common code in another fix.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt [hunk 4]
```diff
@@ -190,71 +275,84 @@ abstract class KotlinIrLinker(
...
                 deserializedTopLevels.add(key)
                 continue
             }
 
-            deserializedModuleDescriptor = key.moduleOfOrigin
             val reachable = deserializeTopLevelDeclaration(key)
             val file = reversedFileIndex[key]!!
             file.declarations.add(reachable)
             reachable.patchDeclarationParents(file)
+            deserializeFileAnnotationsIfFirstUse(moduleOfOrigin, file)
```
Comment (romanart) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt]: I think would be good to put changes about file annotations in separate commit.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/IrModuleDeserializer.kt (hunk 3 above)
Comment (romanart) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/IrModuleDeserializer.kt]: Whatever you like :)

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/compiler.kt (hunk 1 above)
Comment (alexander-gorshenev) [reply compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/compiler.kt]: fixed, it was a rebase artefact

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt (hunk 4 above)
Comment (alexander-gorshenev) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt]: made a separate commit
